### PR TITLE
Lattice parameters

### PR DESCRIPTION
**Pull request by michalkoziarski**
Discrete version of LatticeParameters environment based on the Grid.

### CODE REVIEW

File: gflownet/envs/grid.py [hunk 1]
```diff
@@ -88,10 +89,11 @@ def get_mask_invalid_actions_forward(self, state=None, done=None):
...
         mask = [False for _ in range(self.policy_output_dim)]
         for idx, a in enumerate(self.action_space[:-1]):
+            child = state.copy()
             for d in a:
-                if state[d] + 1 >= self.length:
-                    mask[idx] = True
-                    break
+                child[d] += 1
+            if any(c >= self.length for c in child):
+                mask[idx] = True
```
Comment (alexhernandezgarcia) [inline gflownet/envs/grid.py:96]: Seems fine but out of curiosity, why was this needed?

File: gflownet/envs/lattice_parameters.py [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+
+LATTICE_SYSTEMS = [
+    CUBIC,
+    HEXAGONAL,
+    MONOCLINIC,
+    ORTHORHOMBIC,
+    RHOMBOHEDRAL,
+    TETRAGONAL,
+    TRICLINIC,
+]
```
Comment (alexhernandezgarcia) [inline gflownet/envs/lattice_parameters.py:28]: How about putting these in `gflownet/utils/crystals/constants.py`?

File: gflownet/envs/lattice_parameters.py [hunk 3]
```diff
@@ -0,0 +1,446 @@
...
+    """
+
+    def __init__(
+        self,
+        lattice_system: str,
+        min_length: float = 1.0,
+        max_length: float = 5.0,
+        min_angle: float = 30.0,
+        max_angle: float = 150.0,
+        grid_size: int = 61,
```
Comment (alexhernandezgarcia) [inline gflownet/envs/lattice_parameters.py:51]: Why such a beautiful default number?

File: gflownet/envs/lattice_parameters.py [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+        self.cell2angle, respectively.
+
+        In the case of lengths the allowed actions are:
+            - increment a by n,
+            - increment b by n,
+            - increment c by n,
+            - increment both a and b by n (required by hexagonal, monoclinic and tetragonal lattice systems,
+                for which a == b =/= c),
+            - increment all a, b and c by n (required by cubic and rhombohedral lattice systems, for which
+                a == b == c).
```
Comment (alexhernandezgarcia) [inline gflownet/envs/lattice_parameters.py:172]: Nice!

File: gflownet/envs/lattice_parameters.py [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+
+        In the case of angles the allowed actions are:
+            - increment alpha by n,
+            - increment beta by n,
+            - increment gamma by n,
+            - increment all alpha, beta and gama by n (required by rhombohedral lattice systems, for which
+                alpha == beta == gamma =/= 90 degrees).
```
Comment (alexhernandezgarcia) [inline gflownet/envs/lattice_parameters.py:179]: Also nice :P

File: gflownet/envs/lattice_parameters.py [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+        (a, b, c), _ = self._unpack_lengths_angles(state)
+
+        if self.lattice_system in [CUBIC, RHOMBOHEDRAL]:
+            return a == b == c
+        elif self.lattice_system in [HEXAGONAL, MONOCLINIC, TETRAGONAL]:
+            return a == b != c
+        elif self.lattice_system in [ORTHORHOMBIC, TRICLINIC]:
+            return a != b and a != c and b != c
+        else:
+            raise NotImplementedError
```
Comment (alexhernandezgarcia) [inline gflownet/envs/lattice_parameters.py:250]: Nice too :)

File: gflownet/envs/lattice_parameters.py [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+            state
+        ) and self._are_intermediate_angles_valid(state)
+
+    def _is_final_state_valid(self, state: List[int]) -> bool:
+        """
+        Helper that checks whether the given state meets final self.lattice_system constraints.
+        """
+        return self._are_final_lengths_valid(state) and self._are_final_angles_valid(
+            state
+        )
```
Comment (alexhernandezgarcia) [inline gflownet/envs/lattice_parameters.py:308]: Nice too :)

File: gflownet/envs/lattice_parameters.py [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+            )
+
+        self.source = [0, 0, 0] + [self.angle2cell[angle] for angle in angles]
+
+    def get_actions_space(self) -> List[Tuple[int]]:
+        """
+        Constructs list with all possible actions, including eos.
+
+        The action is described by a tuple of dimensions (possibly duplicate) that will all be incremented
+        by 1, e.g. (0, 0, 0, 2, 4, 4, 4) would increment the 0th and the 4th dimension by 3, and 2nd by 1.
```
[the remark this answers is not in the document]
Comment (michalkoziarski) [reply gflownet/envs/lattice_parameters.py:149]: Well, the first part of the description is correct, right? Action is a tuple of dimensions, and for each d in action, we increment `state[d]` by 1? So with `(0, 0, 0, 2, 4, 4, 4)`, we increment `state[0]` by 1 three times, `state[2]` one time, and `state[4]` three times? Because now I am confused if I got something wrong.

**Review by alexhernandezgarcia — CHANGES_REQUESTED** [review]
Potential issues with constraints

**Review by alexhernandezgarcia — APPROVED** [review]
All looks good to me now! Thanks for the additional updates!
